src/System.Linq/tests/EnumerableTests.cs

```diff
@@ -0,0 +1,237 @@
+            Assert.Equal(4.0/3, fourByThreeWithNulls.Select<int?, double?>(i => i).Average());
+            Assert.Equal(1.0/3, oneByThreeWithNulls.Select<int?, double?>(i => i).Average());
+            // 4. decimal and decimal?
+            Assert.Equal(1, one.Select<int, decimal>(i => i).Average());
+            Assert.Equal(2, two.Select<int, decimal>(i => i).Average());
+            Assert.Equal(4.0m/3, fourByThree.Select<int, decimal>(i => i).Average());
+            Assert.Equal(1.0m/3, oneByThree.Select<int, decimal>(i => i).Average());
+            Assert.Equal(1, oneWithNulls.Select<int?, decimal?>(i => i).Average());
+            Assert.Equal(2, twoWithNulls.Select<int?, decimal?>(i => i).Average());
+            Assert.Equal(4.0m/3, fourByThreeWithNulls.Select<int?, decimal?>(i => i).Average());
+            Assert.Equal(1.0m/3, oneByThreeWithNulls.Select<int?, decimal?>(i => i).Average());
+
+            // Convert the array to different data structures and repeat the tests to
+            // ensure that the method works with other data structures as well.
+            // 1. List
+            Assert.Equal(1, one.ToList().Average());
+            Assert.Equal(2, two.ToList().Average());
+            Assert.Equal(4.0/3, fourByThree.ToList().Average());
+            Assert.Equal(1.0/3, oneByThree.ToList().Average());
+            // 2. Stack
+            Assert.Equal(1, new Stack<int>(one).Average());
+            Assert.Equal(2, new Stack<int>(two).Average());
+            Assert.Equal(4.0/3, new Stack<int>(fourByThree).Average());
+            Assert.Equal(1.0/3, new Stack<int>(oneByThree).Average());
+            // 3. Queue
+            Assert.Equal(1, new Queue<int>(one).Average());
+            Assert.Equal(2, new Queue<int>(two).Average());
+            Assert.Equal(4.0/3, new Queue<int>(fourByThree).Average());
+            Assert.Equal(1.0/3, new Queue<int>(oneByThree).Average());
+        }
+
+        [Fact]
+        public void All()
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            Assert.True(array.All(i => i > 0));
+            for (var j = 1; j <= 10; j++)
+                Assert.False(array.All(i => i > j));
+        }
+
+        [Fact]
+        public void Any()
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            for (var j = 0; j <= 9; j++)
+                Assert.True(array.Any(i => i > j));
+            Assert.False(array.Any(i => i > 10));
+        }
+
+        private void FindCountAndValidate<T>(IEnumerable<T> enumerable, int expectedCount)
+        {
+            Assert.Equal(expectedCount, enumerable.Count());
+            Assert.Equal(expectedCount, enumerable.ToList().Count());
+            Assert.Equal(expectedCount, enumerable.ToArray().Count());
+            Assert.Equal(expectedCount, new Stack<T>(enumerable).Count());
+            Assert.Equal(expectedCount, new Queue<T>(enumerable).Count());
```

Comment:
No, there wasn't any reason, more than I just wanted to test all different collection types without making any assumption about how they are implemented. OK, I will remove one of them.